pkg/kwok/server/debugging_logs.go

```diff
@@ -89,3 +125,445 @@ func (s *Server) getContainerLogs(request *restful.Request, response *restful.Re
 		return
 	}
 }
+
+func (s *Server) getPodLogs(podName, podNamespace, containerName string) (*internalversion.Log, error) {
+	l, has := slices.Find(s.config.Logs, func(l *internalversion.Logs) bool {
+		return l.Name == podName && l.Namespace == podNamespace
+	})
+	if has {
+		l, found := findLogInLogs(containerName, l.Spec.Logs)
+		if found {
+			return l, nil
+		}
+		return nil, fmt.Errorf("not found log target for container %q in pod %q", containerName, log.KRef(podNamespace, podName))
+	}
+
+	for _, cl := range s.config.ClusterLogs {
+		if !cl.Spec.Selector.Match(podName, podNamespace) {
+			continue
+		}
+
+		log, found := findLogInLogs(containerName, cl.Spec.Logs)
+		if found {
+			return log, nil
+		}
+	}
+	return nil, fmt.Errorf("no logs found for container %q in pod %q", containerName, log.KRef(podNamespace, podName))
+}
+
+func findLogInLogs(containerName string, logs []internalversion.Log) (*internalversion.Log, bool) {
+	var defaultLog *internalversion.Log
+	for i, l := range logs {
+		if len(l.Containers) == 0 && defaultLog == nil {
+			defaultLog = &logs[i]
+			continue
+		}
+		if slices.Contains(l.Containers, containerName) {
+			return &l, true
+		}
+	}
+	return defaultLog, defaultLog != nil
+}
+
+// logMessage is the CRI internal log type.
+type logMessage struct {
+	timestamp time.Time
+	stream    runtimeapi.LogStreamType
+	log       []byte
+}
+
+// reset resets the log to nil.
+func (l *logMessage) reset() {
+	l.timestamp = time.Time{}
+	l.stream = ""
+	l.log = nil
+}
+
+// logOptions is the CRI internal type of all log options.
+type logOptions struct {
+	tail      int64
+	bytes     int64
+	since     time.Time
+	follow    bool
+	timestamp bool
+}
+
+// newLogOptions convert the v1.PodLogOptions to CRI internal logOptions.
+func newLogOptions(apiOpts *corev1.PodLogOptions, now time.Time) *logOptions {
+	opts := &logOptions{
+		tail:      -1, // -1 by default which means read all logs.
+		bytes:     -1, // -1 by default which means read all logs.
+		follow:    apiOpts.Follow,
+		timestamp: apiOpts.Timestamps,
+	}
+	if apiOpts.TailLines != nil {
+		opts.tail = *apiOpts.TailLines
+	}
+	if apiOpts.LimitBytes != nil {
+		opts.bytes = *apiOpts.LimitBytes
+	}
+	if apiOpts.SinceSeconds != nil {
+		opts.since = now.Add(-time.Duration(*apiOpts.SinceSeconds) * time.Second)
+	}
+	if apiOpts.SinceTime != nil && apiOpts.SinceTime.After(opts.since) {
+		opts.since = apiOpts.SinceTime.Time
+	}
+	return opts
+}
+
+func readLogs(ctx context.Context, logInfo *internalversion.Log, opts *logOptions, stdout, stderr io.Writer) error {
+	logger := log.FromContext(ctx)
+
+	logsFilePath := path.Join("/var/components/controller", logInfo.LogsFile)
```

Comment:
I think this part of the logic should be handled in kwokctl, not in the kwok server.

Reply:
Working on it
Thank you for reviews a lot! I wanted to push this to ask for an opinion

I'll have it ready now that I know everything how to structure it, and I'll test all scenarios
Thank you again!